components/web_discovery/browser/patterns.cc

```diff
@@ -0,0 +1,321 @@
+               "mapping, or query templates";
+    return std::nullopt;
+  }
+
+  std::vector<PatternsURLDetails> result(url_patterns_list->size());
+
+  for (size_t i = 0; i < url_patterns_list->size(); i++) {
+    auto* url_regex = (*url_patterns_list)[i].GetIfString();
+    if (!url_regex) {
+      VLOG(1) << "URL pattern is not string";
+      return std::nullopt;
+    }
+    auto& details = result[i];
+
+    details.url_regex = std::make_unique<re2::RE2>(*url_regex);
+
+    std::string i_str = base::NumberToString(i);
+
+    auto* id = id_mapping_dict->FindString(i_str);
+    auto* scrape_url_dict = scrape_dict->FindDict(i_str);
+    auto* payloads_url_dict = payloads_dict->FindDict(i_str);
+    if (!id || !scrape_url_dict) {
+      VLOG(1) << "ID or scrape dict missing for pattern";
+      return std::nullopt;
+    }
+    details.id = *id;
+
+    details.is_search_engine =
+        base::ranges::find(search_engines_list->begin(),
+                           search_engines_list->end(),
+                           i_str) != search_engines_list->end();
+
+    auto scrape_rule_groups = ParseScrapeRules(scrape_url_dict);
+    if (!scrape_rule_groups) {
+      return std::nullopt;
+    }
+    details.scrape_rule_groups = std::move(*scrape_rule_groups);
+    if (payloads_url_dict) {
+      auto payload_rule_groups = ParsePayloadRules(payloads_url_dict);
+      if (!payload_rule_groups) {
+        return std::nullopt;
+      }
+      details.payload_rule_groups = std::move(*payload_rule_groups);
+    }
+    if (query_templates_dict) {
+      auto* query_template_dict = query_templates_dict->FindDict(i_str);
+      if (query_template_dict) {
+        auto* prefix = query_template_dict->FindString(kQueryTemplatePrefixKey);
+        if (prefix) {
+          details.search_template_prefix = *prefix;
+        }
+      }
+    }
+  }
+
+  return result;
+}
+
+}  // namespace
+
+ScrapeRule::ScrapeRule() = default;
+ScrapeRule::~ScrapeRule() = default;
+
+PayloadRule::PayloadRule() = default;
+PayloadRule::~PayloadRule() = default;
+
+PayloadRuleGroup::PayloadRuleGroup() = default;
+PayloadRuleGroup::~PayloadRuleGroup() = default;
+
+PatternsURLDetails::PatternsURLDetails() = default;
+PatternsURLDetails::~PatternsURLDetails() = default;
+
+PatternsGroup::PatternsGroup() = default;
+PatternsGroup::~PatternsGroup() = default;
+
+const PatternsURLDetails* PatternsGroup::GetMatchingURLPattern(
+    const GURL& url,
+    bool is_strict_scrape) const {
+  const auto& patterns = is_strict_scrape ? strict_patterns : normal_patterns;
+  for (const auto& pattern : patterns) {
+    if (re2::RE2::PartialMatch(url.spec(), *pattern.url_regex) &&
+        !pattern.scrape_rule_groups.empty()) {
+      return &pattern;
+    }
+  }
+  return nullptr;
+}
+
+std::unique_ptr<PatternsGroup> ParsePatterns(const std::string& patterns_json) {
+  auto result = std::make_unique<PatternsGroup>();
+  auto patterns_value = base::JSONReader::Read(patterns_json);
```

Comment:
where is `patterns_json` coming from? if it's untrusted this should follow the rule of 2.

Reply:
It comes from `patterns.wdp.brave.com` which is trusted

Reply:
@DJAndries how are the patterns generated though? if there's any way an untrusted party could get data into this file it becomes untrusted.

Reply:
The patterns are handwritten and hosted on a CDN controlled by Brave. We depend on TLS for authenticity.

Reply:
[Pinned TLS](https://github.com/brave/brave-core/blob/9299da73ea97b8de4137940fe18df0b3d308cdef/chromium_src/net/tools/transport_security_state_generator/input_file_parsers.cc#L118) to be precise.